File: src/music_orchestrator/providers.py
from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from re import sub

from .config import Settings
from .schemas import Policy, ProviderCapabilities, ProviderRead, ProviderResult, RiskLevel


def canonical_key(title: str, artist: str | None) -> str:
    raw = f"{artist or ''}::{title}".lower()
    return sub(r"[^a-z0-9а-яё]+", " ", raw).strip()


@dataclass(frozen=True)
class TrackSeed:
    provider_track_id: str
    title: str
    artist: str | None = None
    album: str | None = None
    duration_seconds: int | None = None
    source_url: str | None = None
    artwork_url: str | None = None
    attribution: str | None = None


class ProviderAdapter(ABC):
    provider: ProviderRead

    @abstractmethod
    def search(self, query: str, limit: int) -> list[ProviderResult]: ...

    @abstractmethod
    def get(self, provider_track_id: str) -> ProviderResult | None: ...


class LocalProvider(ProviderAdapter):
    def __init__(self, settings: Settings):
        self.settings = settings
        self.provider = ProviderRead(
            id="local",
            name="Local / Navidrome-compatible library",
            kind="local",
            enabled=True,
            configured=bool(settings.navidrome_base_url) or True,
            risk_level=RiskLevel.COMPLIANT,
            capabilities=ProviderCapabilities(
                raw_audio_stream=True,
                persistent_cache=True,
                offline_playback=True,
                public_deployment_safe=True,
            ),
            policy=Policy(
                cache_allowed=True,
                download_allowed=True,
                notes=["Local files only. Navidrome/Subsonic integration is an adapter boundary for future real credentials."],
            ),
            docs_url="https://www.navidrome.org/docs/developers/subsonic-api/",
        )
        self._seeds = [
            TrackSeed("seed-1", "Demo Song", "Local Artist", "Home Library", 180, "/media/demo-song.opus"),
            TrackSeed("seed-1-alt", "Demo Song", "Local Artist", "Home Library", 181, "/media/demo-song-alt.opus"),
            TrackSeed("seed-2", "Another Local Track", "Local Artist", "Home Library", 210, "/media/another-track.opus"),
        ]

    def _to_result(self, seed: TrackSeed) -> ProviderResult:
        return ProviderResult(
            provider_id="local",
            provider_track_id=seed.provider_track_id,
            title=seed.title,
            artist=seed.artist,
            album=seed.album,
            duration_seconds=seed.duration_seconds,
            artwork_url=seed.artwork_url,
            source_url=seed.source_url,
            attribution="Local library",
            risk_level=self.provider.risk_level,
            capabilities=self.provider.capabilities,
            policy=self.provider.policy,
        )

    def search(self, query: str, limit: int) -> list[ProviderResult]:
        q = query.lower().strip()
        matches = [s for s in self._seeds if q in f"{s.title} {s.artist or ''} {s.album or ''}".lower()]
        return [self._to_result(seed) for seed in matches[:limit]]

    def get(self, provider_track_id: str) -> ProviderResult | None:
        return next((self._to_result(seed) for seed in self._seeds if seed.provider_track_id == provider_track_id), None)


class YouTubeOfficialProvider(ProviderAdapter):
    def __init__(self, settings: Settings):
        configured = bool(settings.youtube_api_key)
        self.provider = ProviderRead(
            id="youtube_official",
            name="YouTube official metadata/embed",
            kind="youtube",
            enabled=configured,
            configured=configured,
            risky_enabled=False,
            risk_level=RiskLevel.COMPLIANT,
            capabilities=ProviderCapabilities(
                raw_audio_stream=False,
                embed_playback=True,
                persistent_cache=False,
                offline_playback=False,
                public_deployment_safe=True,
            ),
            policy=Policy(
                cache_allowed=False,
                download_allowed=False,
                requires_external_credentials=True,
                notes=["Metadata/search requires YouTube Data API key. Playback is embed-only. No raw audio, yt-dlp, cache or bypass implementation."],
            ),
            docs_url="https://developers.google.com/youtube/v3",
        )

    def search(self, query: str, limit: int) -> list[ProviderResult]:
        return []

    def get(self, provider_track_id: str) -> ProviderResult | None:
        return ProviderResult(
            provider_id="youtube_official",
            provider_track_id=provider_track_id,
            title="YouTube embedded item",
            artist=None,
            source_url=f"https://www.youtube.com/watch?v={provider_track_id}",
            attribution="YouTube embed playback; metadata not resolved without API key in this MVP.",
            risk_level=self.provider.risk_level,
            capabilities=self.provider.capabilities,
            policy=self.provider.policy,
        )


class SoundCloudOfficialProvider(ProviderAdapter):
    def __init__(self, settings: Settings):
        configured = bool(settings.soundcloud_client_id)
        self.provider = ProviderRead(
            id="soundcloud_official",
            name="SoundCloud official constrained",
            kind="soundcloud",
            enabled=configured,
            configured=configured,
            risky_enabled=False,
            risk_level=RiskLevel.CONSTRAINED,
            capabilities=ProviderCapabilities(
                raw_audio_stream=False,
                official_stream_url=True,
                persistent_cache=False,
                offline_playback=False,
                public_deployment_safe=False,
            ),
            policy=Policy(
                cache_allowed=False,
                download_allowed=False,
                requires_attribution=True,
                requires_external_credentials=True,
                notes=["Official API only; no persistent cache/offline. Custom player must preserve attribution and SoundCloud terms."],
            ),
            docs_url="https://developers.soundcloud.com/docs/api/guide",
        )

    def search(self, query: str, limit: int) -> list[ProviderResult]:
        return []

    def get(self, provider_track_id: str) -> ProviderResult | None:
        return None


def build_adapters(settings: Settings) -> dict[str, ProviderAdapter]:
    adapters = [LocalProvider(settings), YouTubeOfficialProvider(settings), SoundCloudOfficialProvider(settings)]
    return {adapter.provider.id: adapter for adapter in adapters}

File: providers.go
package main

import (
	"encoding/base64"
	"fmt"
	"log/slog"
	"net/url"
	"strings"
	"sync"
)

type ProviderService struct {
	cfg       Config
	extractor *Extractor
}

func NewProviderService(cfg Config) *ProviderService {
	return &ProviderService{cfg: cfg, extractor: NewExtractor(cfg)}
}

func (p *ProviderService) Providers() []Provider {
	return []Provider{youtubeOfficialProvider(), soundcloudOfficialProvider(), p.extractorProvider("youtube_stream", "YouTube Stream", "youtube"), p.extractorProvider("soundcloud_stream", "SoundCloud Stream", "soundcloud")}
}
func (p *ProviderService) byID(id string) (Provider, bool) {
	for _, pr := range p.Providers() {
		if pr.ID == id {
			return pr, true
		}
	}
	return Provider{}, false
}
func localCaps() Capabilities {
	return Capabilities{Search: true, Metadata: true, RawAudioStream: true, PersistentCache: true, OfflinePlayback: true, PublicDeploymentSafe: true}
}
func localPolicy() Policy {
	return Policy{DownloadAllowed: true, CacheAllowed: true, Notes: []string{"Local/demo media is safe to stream and cache."}}
}
func localProvider() Provider {
	return Provider{ID: "local", Name: "Local Demo Library", Kind: "local", Configured: true, Enabled: true, RiskLevel: "compliant", Capabilities: localCaps(), Policy: localPolicy()}
}
func youtubeOfficialProvider() Provider {
	return Provider{ID: "youtube_official", Name: "YouTube Official Embed", Kind: "youtube", Configured: false, Enabled: false, RiskLevel: "constrained", Capabilities: Capabilities{Search: false, Metadata: true, EmbedPlayback: true, PublicDeploymentSafe: true}, Policy: Policy{DownloadAllowed: false, CacheAllowed: false, RequiresAttribution: true, Notes: []string{"Use official embeds/API only."}}}
}
func soundcloudOfficialProvider() Provider {
	return Provider{ID: "soundcloud_official", Name: "SoundCloud Official", Kind: "soundcloud", Configured: false, Enabled: false, RiskLevel: "constrained", Capabilities: Capabilities{Search: false, Metadata: true, EmbedPlayback: true, PublicDeploymentSafe: true}, Policy: Policy{DownloadAllowed: false, CacheAllowed: false, RequiresAttribution: true, Notes: []string{"Official API credentials are optional future work."}}}
}
func (p *ProviderService) extractorProvider(id, name, kind string) Provider {
	return Provider{ID: id, Name: name, Kind: kind, Configured: p.cfg.EnableRiskyExtractors, Enabled: p.cfg.EnableRiskyExtractors, RiskLevel: "risky", Capabilities: Capabilities{Search: p.cfg.EnableRiskyExtractors, Metadata: p.cfg.EnableRiskyExtractors, RawAudioStream: p.cfg.EnableRiskyExtractors, PersistentCache: p.cfg.EnableRiskyExtractors, OfflinePlayback: p.cfg.EnableRiskyExtractors, PublicDeploymentSafe: false}, Policy: Policy{DownloadAllowed: p.cfg.EnableRiskyExtractors, CacheAllowed: p.cfg.EnableRiskyExtractors, RequiresAttribution: true, Notes: []string{"Personal self-hosted mode only. Users are responsible for source terms and law."}}}
}

func (p *ProviderService) Search(query string, providerIDs []string, limit int) []Track {
	if len(providerIDs) == 0 {
		if p.cfg.EnableRiskyExtractors {
			providerIDs = []string{"youtube_stream", "soundcloud_stream"}
		}
	}
	extractorIDs := []string{}
	for _, id := range providerIDs {
		switch id {
		case "youtube_stream", "soundcloud_stream":
			if p.cfg.EnableRiskyExtractors {
				extractorIDs = append(extractorIDs, id)
			}
		}
	}
	// Each extractor search is a cold yt-dlp process, so running them one after
	// another made total latency the sum of both. Fan out and keep the results
	// in request order so the provider ordering the caller asked for survives.
	results := make([][]Track, len(extractorIDs))
	var wg sync.WaitGroup
	for i, id := range extractorIDs {
		wg.Add(1)
		go func(idx int, providerID string) {
			defer wg.Done()
			items, err := p.extractor.Search(providerID, query, limit)
			if err != nil {
				slog.Warn("provider search failed", "provider", providerID, "error", err)
				return
			}
			results[idx] = items
		}(i, id)
	}
	wg.Wait()
	out := []Track{}
	for _, items := range results {
		if limit > 0 && len(out) >= limit {
			break
		}
		out = append(out, items...)
	}
	if limit > 0 && len(out) > limit {
		return out[:limit]
	}
	return out
}
func localSearch(query string) []Track {
	q := strings.ToLower(query)
	seeds := []Track{localTrack("seed-1", "Demo Song One", "Local Artist", 183), localTrack("seed-2", "Demo Lofi Loop", "Local Artist", 122), localTrack("seed-3", "Demo Synthwave Night", "Local Artist", 201)}
	if q == "" {
		return seeds
	}
	out := []Track{}
	for _, t := range seeds {
		if strings.Contains(strings.ToLower(t.Title+" "+t.Artist), q) {
			out = append(out, t)
		}
	}
	return out
}
func localTrack(pid, title, artist string, dur int) Track {
	return Track{ID: "local:" + pid, ProviderID: "local", ProviderTrackID: pid, Title: title, Artist: artist, DurationSeconds: dur, SourceURL: "/media/demo-" + pid + ".mp3", Attribution: "Local demo", Capabilities: localCaps(), Policy: localPolicy()}
}
func (p *ProviderService) Track(trackID string) (Track, bool) {
	provider, pid, ok := splitTrackID(trackID)
	if !ok {
		return Track{}, false
	}
	switch provider {
	case "local":
		for _, t := range localSearch("") {
			if t.ProviderTrackID == pid {
				return t, true
			}
		}
	case "youtube_stream", "soundcloud_stream":
		if p.cfg.EnableRiskyExtractors {
			t, err := p.extractor.Resolve(provider, pid)
			return t, err == nil
		}
	}
	return Track{}, false
}
func splitTrackID(id string) (string, string, bool) {
	a, b, ok := strings.Cut(id, ":")
	return a, b, ok && a != "" && b != ""
}
func ytURL(pid string) string     { return "https://www.youtube.com/watch?v=" + url.QueryEscape(pid) }
func scIDFromURL(u string) string { return "url_" + base64.RawURLEncoding.EncodeToString([]byte(u)) }
func scURLFromID(id string) (string, error) {
	raw := id
	if strings.HasPrefix(id, "url_") {
		b, err := base64.RawURLEncoding.DecodeString(strings.TrimPrefix(id, "url_"))
		if err != nil {
			return "", err
		}
		raw = string(b)
	}
	u, err := url.Parse(raw)
	if err != nil || (u.Scheme != "https" && u.Scheme != "http") || u.Hostname() == "" {
		return "", fmt.Errorf("soundcloud extractor id must be a valid URL")
	}
	host := strings.ToLower(u.Hostname())
	if host != "soundcloud.com" && !strings.HasSuffix(host, ".soundcloud.com") && host != "sndcdn.com" && !strings.HasSuffix(host, ".sndcdn.com") {
		return "", fmt.Errorf("soundcloud URL host is not allowed")
	}
	return u.String(), nil
}
